sound/soc/sof/sof-audio.c

```diff
@@ -350,13 +349,30 @@ int sof_widget_list_setup(struct snd_sof_dev *sdev, struct snd_sof_pcm *spcm, in
 	/* set up widgets in the list */
 	for_each_dapm_widgets(list, num_widgets, widget) {
 		struct snd_sof_widget *swidget = widget->dobj.private;
+		struct snd_sof_widget *pipe_widget;
 
 		if (!swidget)
 			continue;
 
-		ret = sof_widget_setup(sdev, swidget);
+		/* set up the pipeline widget */
+		pipe_widget = swidget->pipe_widget;
+		if (!pipe_widget) {
+			dev_err(sdev->dev, "error: no pipeline widget found for %s\n",
+				swidget->widget->name);
+			ret = -EINVAL;
+			goto widget_free;
+		}
+
+		ret = sof_widget_setup(sdev, pipe_widget);
```

Comment:
I don't fully get the idea here: 
Assuming all the widgets "belong" to a pipeline, are we really going to set-up the pipeline multiple times?
Shouldn't we do so only once?

Reply:
sure, done!